src/resource.rs

```diff
@@ -688,3 +688,48 @@ impl FromIterator<ResourceId> for ResourceSet {
         Self(HashSet::from_iter(iter))
     }
 }
+
+#[cfg(test)]
+mod test {
+    use crate::types::{
+        type_param::{TypeArg, TypeParam},
+        ClassicType, HashableType, PrimType, SimpleType, TypeTag,
+    };
+
+    use super::{TypeDef, TypeDefTag};
+
+    #[test]
+    fn test_instantiate_typedef() {
+        let def = TypeDef {
+            name: "MyType".into(),
+            params: vec![TypeParam::ClassicType],
+            resource: Some("MyRsrc".into()),
+            description: "Some parameterised type".into(),
+            tag: TypeDefTag::FromParams(vec![0]),
+        };
+        let typ: SimpleType = def
+            .instantiate_concrete(vec![TypeArg::ClassicType(ClassicType::F64)])
+            .unwrap()
+            .into();
+        assert_eq!(typ.tag(), TypeTag::Classic);
+        let typ2: SimpleType = def
+            .instantiate_concrete([TypeArg::ClassicType(ClassicType::Hashable(
+                HashableType::String,
+            ))])
+            .unwrap()
+            .into();
+        assert_eq!(typ2.tag(), TypeTag::Hashable);
+
+        // And some bad arguments...firstly, wrong kind of TypeArg:
+        def.instantiate_concrete([TypeArg::HashableType(HashableType::String)])
+            .unwrap_err();
+        // Too few arguments:
+        def.instantiate_concrete([]).unwrap_err();
```

Comment:
This one was previously panicking with an index-out-of-bounds in computing the tag

Reply:
(If I'd provided a TypeDefTag::Explicit it would have just returned a CustomType)